File: examples/js/nebula-exploration.ts
/**
 * Example: scan a nebula from a dApp (module: nebula_explorer).
 *
 *   NEBULA_CONTRACT_ID=C... npx tsx nebula-exploration.ts
 *
 * Steps:
 *   1. preview a layout with a free simulation (`generate_nebula_layout`),
 *   2. submit `scan_nebula`, which generates the layout, scores its rarity
 *      (same logic as `calculate_rarity_tier`) and emits the event indexers
 *      and the frontend listen for.
 */
import { arg, config, invoke, loadPlayer, randomSeed, requireContractId, view } from "./lib/contract";

interface NebulaCell {
  x: number;
  y: number;
  cell_type: unknown;
  energy: number;
}
interface NebulaLayout {
  width: number;
  height: number;
  cells: NebulaCell[];
  seed: Buffer;
  timestamp: bigint;
  total_energy: number;
}

async function main() {
  requireContractId(config.contractId, "NEBULA_CONTRACT_ID");
  const player = await loadPlayer();

  // A fresh 32-byte seed. The contract mixes in ledger data, so the result
  // depends on the ledger the transaction lands in.
  const seed = randomSeed();

  // Step 1: free preview via simulation (no fee, no state change).
  const preview = await view<NebulaLayout>(player, "generate_nebula_layout", [
    arg.bytes32(seed),
    arg.address(player.publicKey()),
  ]);
  console.log(`Preview: ${preview.width}x${preview.height}, energy ${preview.total_energy}`);

  // Step 2: submit the scan. Returns (NebulaLayout, Rarity) as a 2-tuple.
  const [layout, rarity] = await invoke<[NebulaLayout, unknown]>(player, "scan_nebula", [
    arg.bytes32(seed),
    arg.address(player.publicKey()),
  ]);
  const rarityName = Array.isArray(rarity) ? rarity[0] : rarity; // enums decode as ["Variant"]
  console.log(`Scanned: energy ${layout.total_energy}, rarity ${String(rarityName)}`);

  // Show the richest cells so the UI can highlight them.
  const top = [...layout.cells].sort((a, b) => b.energy - a.energy).slice(0, 5);
  for (const c of top) {
    console.log(`  (${c.x},${c.y}) energy=${c.energy}`);
  }
}

main().catch((e) => {
  console.error(e);
  process.exit(1);
});

File: examples/nebula_exploration.rs
//! # Example: Nebula exploration (main contract)
//!
//! Module: `nebula_explorer` via `NebulaNomadContract`
//!
//! Shows the core exploration loop:
//!   1. generate a procedural 16x16 nebula from a 32-byte seed,
//!   2. score its rarity,
//!   3. do both in one call with `scan_nebula`.
//!
//! Run:
//! ```text
//! cargo run --example nebula_exploration
//! ```

use soroban_sdk::{testutils::Address as _, Address, BytesN, Env};
use stellar_nebula_nomad::{NebulaNomadContract, NebulaNomadContractClient};

fn main() {
    // ── Step 1: Local Soroban environment ────────────────────────────────
    // `Env::default()` is an in-memory ledger. `mock_all_auths()` makes every
    // `require_auth()` pass so we can focus on game logic. On a real network
    // the player's wallet signs instead.
    let env = Env::default();
    env.mock_all_auths();

    // ── Step 2: Deploy the contract and create a typed client ────────────
    let contract_id = env.register(NebulaNomadContract, ());
    let client = NebulaNomadContractClient::new(&env, &contract_id);

    let player = Address::generate(&env);

    // ── Step 3: Pick a seed ──────────────────────────────────────────────
    // Any 32 bytes work. The contract mixes in the ledger sequence and
    // timestamp, so the same seed gives a different nebula on each ledger.
    // dApps typically use `crypto.getRandomValues`.
    let seed = BytesN::from_array(&env, &[7u8; 32]);

    // ── Step 4: Generate a layout ────────────────────────────────────────
    let layout = client.generate_nebula_layout(&seed, &player);
    println!("Generated nebula:");
    println!("  size          : {}x{}", layout.width, layout.height);
    println!("  cells         : {}", layout.cells.len());
    println!("  total energy  : {}", layout.total_energy);
    println!("  generated at  : {}", layout.timestamp);

    // ── Step 5: Score its rarity ─────────────────────────────────────────
    // Read-only helper: rare cells * 10 + energy density -> tier.
    let rarity = client.calculate_rarity_tier(&layout);
    println!("  rarity tier   : {rarity:?}");

    // ── Step 6: One-shot scan (generate + score + event) ─────────────────
    // `scan_nebula` is what a game client normally calls. It also emits the
    // `NebulaScanned` event that indexers and the frontend listen for.
    let (scanned, scanned_rarity) = client.scan_nebula(&seed, &player);
    println!();
    println!("scan_nebula:");
    println!("  energy={} rarity={scanned_rarity:?}", scanned.total_energy);

    // ── Step 7: Determinism ─────────────────────────────────────────────
    // Same seed + same ledger = same layout, so anyone can verify a scan.
    let again = client.generate_nebula_layout(&seed, &player);
    assert_eq!(again.total_energy, layout.total_energy);
    println!();
    println!("Determinism check passed: identical inputs -> identical layout.");
}
